### PR TITLE
feat(inputs.opcua): Implement browsing tags on initial connect

### LINKED ISSUE #15001: Add a way to resolve OPC-UA tags by name when the server only permits numeric identifiers
### Use Case
I would like to propose to implement a method to resolve and subscribe to OPC-UA nodes by BrowseName when the server only permits numeric identifiers.

Issues may arise when using numeric identifiers and the OPC-UA server configuration changes. The server-side identifiers may change and if the configuration on the client side is not changed accordingly, the wrong tags may be logged. Some OPC-UA servers will only respond to numeric identifiers so the setup may become fragile.

### Expected behavior

The capacity to add a tag by browseable and have the server resolve it to a numeric identifier on establishing connection by using the BrowseName node attributes.

This may be done by adding a new identifier_type in the config, such as `r` (to be determined), which will make the server first resolve the tag symbolically using the BrowseName attributes of the server nodes after the connection is established. 
Example:
`{name="Some.Hierarchical.Tag", namespace="1", identifier_type="r", identifier="Interface:Some.Hierarchical.Identifier"},`

### Actual behavior

The OPC-UA client can only use the default OPC-UA identifier types (string, numeric, guid, opaque). The capacity to resolve a numeric identifier by using the BrowseName fields is non-existing.

### Additional info
I'm currently experiencing this issue with a Siemens S7-1200 right now which has this issue.
I've started working on a proof of concept for this implementation myself because this problem repeatedly breaks my setup. Because I'm not very well versed in OPC-UA, Telegraf and have only a little experience Go, and this is pretty much my first open-source contribution, I could use some guidance in this.

### PR DESCRIPTION
**Pull request by supergrover**
## Summary
This is a proposed implementation in order to implement tag browsing for OPC-UA tags on initial connect as described in #15001.
This pull request is for review purposes, some tests are still failing on this because a connection to an OPC server is now required in order initialize all NodeIDs. I'm still working on that part.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #15001

### CODE REVIEW

File: plugins/common/opcua/input/input_client.go [hunk 1]
```diff
@@ -57,6 +58,7 @@ type NodeSettings struct {
 	TagsSlice        [][]string           `toml:"tags" deprecated:"1.25.0;use 'default_tags' instead"`
 	DefaultTags      map[string]string    `toml:"default_tags"`
 	MonitoringParams MonitoringParameters `toml:"monitoring_params"`
+	Browse           bool                 `toml:"browse"`
```
Comment (srebhan) [inline plugins/common/opcua/input/input_client.go:61]: How about making this a pointer like
```suggestion
	Browse           *bool                `toml:"browse"`
```
as this allows you to decide whether the value was actually _set_ by the user or not.

File: plugins/common/opcua/input/input_client.go [hunk 2]
```diff
@@ -352,6 +355,9 @@ func (o *OpcUAInputClient) InitNodeMetricMapping() error {
 			if node.MonitoringParams.SamplingInterval == 0 {
 				node.MonitoringParams.SamplingInterval = group.SamplingInterval
 			}
+			if node.Browse == false {
+				node.Browse = group.Browse
+			}
```
Comment (srebhan) [inline plugins/common/opcua/input/input_client.go:360]: With the above change you can do
```suggestion
			if node.Browse == nil {
				node.Browse = &group.Browse
			}
```

File: plugins/common/opcua/input/input_client.go [hunk 3]
```diff
@@ -368,19 +374,111 @@ func (o *OpcUAInputClient) InitNodeMetricMapping() error {
...
+			} else {
+				nid = browsedNodeID
+			}
+		} else {
+			parsedNodeID, err := ua.ParseNodeID(node.Tag.NodeID())
+			if err != nil {
+				return err
+			}
+			nid = parsedNodeID
 		}
```
Comment (srebhan) [inline plugins/common/opcua/input/input_client.go:403]: How about
```suggestion
		var nid *ua.NodeID = nil

		if node.Tag.Browse && node.Tag.IdentifierType == "s" {
			browsedNodeID, err := o.browseNode(ctx, node.Tag.Namespace, node.Tag.Identifier)
			if err == nil {
				o.NodeIDs = append(o.NodeIDs, browsedNodeID)
				continue
			}
			
				// Continue as other nodes might still be browseable. 
				// Add the node as a string identifier, the server will not be able to return data
				// but we must add a node to the o.NodeIDs array in order for it to match the
				// o.NodeMetricMappings
				o.Log.Errorf("Browsing node %q failed: %v", node.Tag.FieldName, err)
		} 
		parsedNodeID, err := ua.ParseNodeID(node.Tag.NodeID())
		if err != nil {
			return err
		}
		o.NodeIDs = append(o.NodeIDs, parsedNodeID)
```

File: plugins/common/opcua/input/input_client.go [hunk 4]
```diff
@@ -368,19 +374,111 @@ func (o *OpcUAInputClient) InitNodeMetricMapping() error {
...
 }
 
+func parseNamespaceId(ns string) (uint16, error) {
+	nsId, err := strconv.ParseUint(ns, 10, 16)
+	if err != nil {
+		return 0, fmt.Errorf("invalid namespace number: %s", err)
+	}
+
+	return uint16(nsId), nil
+}
```
Comment (srebhan) [inline plugins/common/opcua/input/input_client.go:418]: Please merge this into the respective calling places as this is not worth a wrapper.

File: plugins/common/opcua/input/input_client.go [hunk 5]
```diff
@@ -368,19 +374,111 @@ func (o *OpcUAInputClient) InitNodeMetricMapping() error {
...
+			NamespaceIndex: segmentNs,
+			Name:           segmentName,
+		})
+	}
+
+	return pathNames, nil
+}
+
+func (o *OpcUAInputClient) browseNode(ctx context.Context, ns string, browsePath string) (*ua.NodeID, error) {
+	o.Log.Infof("Browsing node %s", browsePath)
```
Comment (srebhan) [inline plugins/common/opcua/input/input_client.go:465]: This should at most be a debug log.

File: plugins/inputs/opcua_listener/opcua_listener_test.go [hunk 6]
```diff
@@ -800,6 +800,7 @@ func TestSubscribeClientConfigValidMonitoringParams(t *testing.T) {
 	})
 
 	subClient, err := subscribeConfig.CreateSubscribeClient(testutil.Logger{})
+	subClient.Connect()
```
Comment (srebhan) [inline plugins/inputs/opcua_listener/opcua_listener_test.go:803]: Why do you need to modify this test?

File: plugins/inputs/opcua_listener/subscribe_client.go [hunk 7]
```diff
@@ -124,6 +110,22 @@ func (o *SubscribeClient) Connect() error {
 		return err
 	}
 
+	// Make sure we setup the node-ids correctly after reconnect
+	// as the server might be restarted and IDs changed
+	o.Log.Debugf("Creating monitored items on first connection")
```
Comment (srebhan) [inline plugins/inputs/opcua_listener/subscribe_client.go:115]: This log line does not help anyone. ;-)

**Review by srebhan — COMMENTED** [review]
@supergrover first of all thank you for your pull-request! I do have some comments in the code and additionally the following more high-level comments:
1. Please run `make lint` to see the linter issues in your code. Please fix them!
2. How does this code relate to [gopcua's BrowseRequest](https://pkg.go.dev/github.com/gopcua/opcua@v0.5.3/ua#BrowseRequest)? Can we make use of the upstream functionality instead of coding this by hand?

**Comment by supergrover** [comment]
Thanks for the suggestions! I'll work on them and get back to this.

>     1. Please run `make lint` to see the linter issues in your code. Please fix them!
Sure, I did not know about the linter until Github started screaming at me :)
>     2. How does this code relate to [gopcua's BrowseRequest](https://pkg.go.dev/github.com/gopcua/opcua@v0.5.3/ua#BrowseRequest)? Can we make use of the upstream functionality instead of coding this by hand?
The current code basically parses the string formatted browse path and passes it to the [opcua.Node.TranslateBrowsePathsToNodeIDs](https://pkg.go.dev/github.com/gopcua/opcua@v0.5.3#Node.TranslateBrowsePathsToNodeIDs) method on the gopcua client, so basically this already is an upstream function we're calling. Although the parsing of the browse paths could be a suggestion for upstream, if that's what you mean.
The `BrowseRequest` you mentioned confused me for a while, but as far as I understand it's intended for actual browsing (i.e. listing all subnodes of a node which satisfy some criteria) instead of asking the server for the Node which corresponds to a BrowsePath.

> Why do you need to modify this test?

That's part of my work on seeing what needs to be done to satisfy the tests which appears to have slipped through. Basically, the previous code parsed all the node configurations and created the monitored item requests on creation of the subscribe client in [https://github.com/influxdata/telegraf/blob/master/plugins/inputs/opcua_listener/subscribe_client.go#L111](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/opcua_listener/subscribe_client.go#L111).
Because nodes may need to be browsed, we cannot know the NodeID's them until we have connected to an OPC server in order to ask it for the NodeID's belonging to the BrowsePath.

**Comment by srebhan** [comment]
> Sure, I did not know about the linter until Github started screaming at me :)

No worries! It would be good to fix them before pushing so we save some CircleCI credits. ;-)

> > 2. How does this code relate to [gopcua's BrowseRequest](https://pkg.go.dev/github.com/gopcua/opcua@v0.5.3/ua#BrowseRequest)? Can we make use of the upstream functionality instead of coding this by hand?
> 
> The current code basically parses the string formatted browse path and passes it to the [opcua.Node.TranslateBrowsePathsToNodeIDs](https://pkg.go.dev/github.com/gopcua/opcua@v0.5.3#Node.TranslateBrowsePathsToNodeIDs) method on the gopcua client, so basically this already is an upstream function we're calling. Although the parsing of the browse paths could be a suggestion for upstream, if that's what you mean. The `BrowseRequest` you mentioned confused me for a while, but as far as I understand it's intended for actual browsing (i.e. listing all subnodes of a node which satisfy some criteria) instead of asking the server for the Node which corresponds to a BrowsePath.

Thanks for your explanation! Yeah I wondered why you manually need to mangle the path and suspected an upstream function for doing this. :-) I'm OK with doing it in the plugin but if you want to maybe suggesting this upstream is also a good thing. We can still switch over to it later...

> > Why do you need to modify this test?
> 
> That's part of my work on seeing what needs to be done to satisfy the tests which appears to have slipped through. Basically, the previous code parsed all the node configurations and created the monitored item requests on creation of the subscribe client in https://github.com/influxdata/telegraf/blob/master/plugins/inputs/opcua_listener/subscribe_client.go#L111. Because nodes may need to be browsed, we cannot know the NodeID's them until we have connected to an OPC server in order to ask it for the NodeID's belonging to the BrowsePath.

I see, then please check _all_ errors before continuing! I think the `Connect()` needs to be further down and also the error returned by `Connect()` need to be checked in the test!

**Pull request closed without merging**